src/main/java/org/spongepowered/api/text/transform/TextFormatter.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.text.transform;
+
+import com.google.common.collect.ImmutableList;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.TextRepresentable;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * A TextFormatter is a mutable collection of {@link TextRepresentable}s which
+ * are all concatenated to an empty {@link Text} object on {@link #format()}.
+ */
+public interface TextFormatter<E extends TextRepresentable> extends TextRepresentable, Iterable<E> {
+
+    /**
+     * Returns an {@link ImmutableList} of this formatter's elements.
+     *
+     * @return All elements
+     */
+    ImmutableList<E> getAll();
+
+    /**
+     * Returns the element at the specified index.
+     *
+     * @param i Index to retrieve from
+     * @return Element at index
+     */
+    default E get(int i) {
+        return getAll().get(i);
```

Comment:
It shouldn't be necessary to make a copy of the whole list before being able to get a single element.